### ISSUE TITLE
Token lifetime and invalidation

### ISSUE THREAD

**Issue by FranCarstens** [patrickpissurno/fastify-esso#18]
This may be an ignorant question, but I'm not seeing any reference to invalidation or expiration in the docs.

I'm expiring tokens in my application, but I'm wondering how, or if tokens become invalid/expire on the server. 

Is this even something I should be concerned about?

**Comment by patrickpissurno** [comment]
Hey! Sorry for not getting back to you sooner.

When you call `fastify.generateAuthToken`, you can pass an object with any values you'd like to. So, by leveraging that, you can do something like this:

```js
const token = await fastify.generateAuthToken({ expires: moment().add(1, 'hour').toDate() });
```

Then, when registering `fastify-esso`, you can use the `extra_validation` option to validate that, like this:

```js
const opts = {
    /** Secure key that will be used to do the encryption stuff */
    secret: '11111111111111111111', // cannot be omitted
    
    /** request and reply are Fastify's request and reply objects **/
    extra_validation: async function validation (request, reply){
        if(moment().isAfter(new Date(req.auth.expires)))
            throw new Error('Your token has expired. Please sign-in again');
    },
};

fastify.register(require('fastify-esso')(opts));
```

You can do much more advanced things with `extra_validation` too.

The only rule is it should throw if validation fails. Otherwise, it shall NOT return anything. This is very important because if it does return something, it would cause the response to be sent earlier to the browser, effectively preventing the actual route code from running.

For these examples, I used [moment.js](https://momentjs.com/) for simplicity, but if you're not using it already you may be better off avoiding it (it's been deprecated) and using something like [date-fns](https://date-fns.org/) instead.

**Comment by patrickpissurno** [comment]
The only thing that's harder and I haven't really figured out a good solution for is blocklists.

Tokens are immutable, so to expire them earlier, you would need to have something external, like a Redis instance, where you could save the ones that were blocked, and then use `extra_validation` to make sure the token is not in that Redis DB.

Note that this is also an issue when using regular JWTs, and having to access something external for every request kind of diminishes the benefits this approach brings. Eg.: you could just store the tokens in that external DB and avoid JWTs (and similars) altogether, except that with this, instead of having to store every token, you just have to store the blocked ones for their maximum duration. So, by doing this you'll still get a much more lightweight storage footprint, which means it's easier/cheaper to use something like Redis (that keeps everything on RAM).

TL;DR: implementing blocklists with this approach comes with tradeoffs.

Hope this helps!
